test/grouping.jl

```diff
@@ -606,6 +631,94 @@ module TestGrouping
         end
     end
 
+    struct TestType end
+    Base.isless(::TestType, ::Int) = true
+    Base.isless(::Int, ::TestType) = false
+    Base.isless(::TestType, ::TestType) = false
+
+    @testset "combine with aggregation functions" begin
+        Random.seed!(1)
```

Comment:
We could run scenarios:
* `df = DataFrame(a = rand(1:5, 20), x1 = rand(Int, 20), x2 = rand(Complex{Int}, 20))`
* `df = DataFrame(a = categorical(rand(1:5, 20)), x1 = rand(Int, 20), x2 = rand(Complex{Int}, 20))`
* `df = DataFrame(a = categorical(rand(1:5, 20)), x1 = rand(Int, 20), x2 = rand(Complex{Int}, 20)); levels!(df.a, shuffle([levels(df.a); 10; 11]))`
* `df = DataFrame(a = categorical(rand(1:5, 20), ordered=true), x1 = rand(Int, 20), x2 = rand(Complex{Int}, 20))`
* `df = DataFrame(a = categorical(rand(1:5, 20), ordered=true), x1 = rand(Int, 20), x2 = rand(Complex{Int}, 20)); levels!(df.a, shuffle([levels(df.a); 10; 11]))`
e.g. 10 times to automatically improve the coverage.

Reply:
Unfortunately that's not so simple. Most operations don't work for categorical values. I could add some cases in the categorical-specific tests, but what do you want to test exactly? For example, unused levels shouldn't make any difference here, and AFAIK they are tested elsewhere already (where it's more relevant). But please point uncovered scenarios if they are missing.

Reply:
My point is that it is a complex design so randomizing tests might catch something we have not thought of. The question is do we fully cover with tests:
* grouping on one column that is categorical ordered and unordered with missing level and non-missing level (these are 4 scenarios);
* the same but when grouping on multiple columns all categorical (possibly with mixed characteristics);
* the same but when grouping on multiple columns some categorical and some not-categorical;

If we have all these scenarios covered I am OK to leave it as is (or "reasonably covered" - i.e. we do not have to test things that you are 100% sure that work the same as the things that are tested).

But in general - that is why I have approved earlier because I think that it is only an extra precaution.

Reply:
OK. What I meant is that to avoid combinatorial explosion of tests, I've tried to test each feature only once. All variations regarding the *grouping* column are handled in tests for `groupby`, since they make no difference for the `combine` code -- and AFAICT empty and reordered levels are tested there already. I guess we could add more tests with variations on the *aggregated* column here, but it's hard to decide what's relevant and what's not...

Reply:
OK - let us leave it as is. I think that aggregated column is less risky.